src/biome/text/api_new/modules/heads/bimpm_classification.py

```diff
@@ -0,0 +1,487 @@
+from typing import Optional, Dict, List, Any
+
+import torch
+from allennlp.common.checks import check_dimensions_match
+from allennlp.data import Instance
+from allennlp.modules import (
+    FeedForward,
+    Seq2SeqEncoder,
+    Seq2VecEncoder,
+    TimeDistributed,
+)
+from allennlp.modules.bimpm_matching import BiMpmMatching
+from allennlp.nn import InitializerApplicator
+from allennlp.nn import util
+from biome.text.api_new.model import Model
+from biome.text.api_new.modules.encoders import TimeDistributedEncoder
+from biome.text.api_new.modules.heads.classification.defs import ClassificationHead
+from overrides import overrides
+
+
+class BiMpm(ClassificationHead):
+    """
+    This ``Model`` is a version of AllenNLPs implementation of the BiMPM model described in
+    `Bilateral Multi-Perspective Matching for Natural Language Sentences <https://arxiv.org/abs/1702.03814>`_
+    by Zhiguo Wang et al., 2017.
+
+    This version adds the feature of being compatible with multiple inputs for the two records.
+    The matching will be done for all possible combinations between the two records, that is:
+    (r1_1, r2_1), (r1_1, r2_2), ..., (r1_2, r2_1), (r1_2, r2_2), ...
+
+    This version also allows you to apply only one encoder, and to leave out the backward matching,
+    providing the possibility to use transformers for the encoding layer.
+
+    Parameters
+    ----------
+    model : ``Model``
+        Takes care of the embedding
+    labels : `List[str]`
+        List of labels
+    matcher_word : ``BiMpmMatching``
+        BiMPM matching on the output of word embeddings of record1 and record2.
+    encoder : ``Seq2SeqEncoder``
```

Comment:
I think we should move this encoder to the backbone model, to make this model more consistent and to allow for reusing encoders as we do for other heads.

Reply:
I tried to move it, but it was messy, since the `matcher_word` compares the embeddings, and therefore needs the ouput of the `model.embedder`. I find defining a `PassThroughEncoder` as the `model.encoder` a much cleaner solution for now.